### PR TITLE
Implementation of `portAbsent` and `outputControlReaction` in federated execution of TypeScript

### PR DESCRIPTION
**Pull request by byeong-gil**
Relevant issue: #105

Relevant LF PR: [lf-lang/lingua-franca#1333](https://github.com/lf-lang/lingua-franca/pull/1333)

Add logic for sending and handling MSG_TYPE_PORT_ABSENT and triggering output control reactions.

Some functions of functions are added and modified like sendRTIPortAbsent(), registerOutputControlReaction() etc.

The logic for handling input control reaction should be added and this will be updated soon.

### CODE REVIEW

File: src/core/federation.ts [hunk 1]
```diff
@@ -828,6 +864,23 @@ class RTIClient extends EventEmitter {
                     bufferIndex += 13;
                     break;
                 }
+                case RTIMessageTypes.MSG_TYPE_PORT_ABSENT: {
+                    // The next 2 bytes is the port id.
```
Comment (hokeun) [inline src/core/federation.ts:868]: NIT: next 2 bytes is -> next 2 bytes are

File: src/core/federation.ts [hunk 2]
```diff
@@ -1178,6 +1254,25 @@ export class FederatedApp extends App {
         this.rtiClient.sendRTIStopRequestReply(tag);
     }
 
+    /**
+     * Send a port absent message to federate with fed_ID, informing the
```
Comment (hokeun) [inline src/core/federation.ts:1258]:
```suggestion
     * Send a port absent message to the destination port, informing the
```

File: src/core/federation.ts [hunk 3]
```diff
@@ -1178,6 +1254,25 @@ export class FederatedApp extends App {
...
+     * remote federate that the current federate will not produce an event
+     * on this network port at the current logical time.
+     * 
+     * @param additional_delay The offset applied to the timestamp
```
Comment (hokeun) [inline src/core/federation.ts:1262]:
```suggestion
     * @param additionalDelay The offset applied to the timestamp
```

File: src/core/federation.ts [hunk 4]
```diff
@@ -1178,6 +1254,25 @@ export class FederatedApp extends App {
...
+     *  using after. The additional delay will be greater or equal to zero
+     *  if an after is used on the connection. If no after is given in the
+     *  program, NEVER is passed.
```
Comment (hokeun) [inline src/core/federation.ts:1265]: `If no after is given in the program, NEVER is passed` -> This statement doesn't seem true from the implementation. Maybe we should update the implementation to match this description?

**Comment by byeong-gil** [comment]
This PR works well except for tests that failed in the `fed-gen` branch. 
We can see reactors send and receive `port-absent` messages every time when the port doesn't make outputs.

File: src/core/reactor.ts [hunk 5]
```diff
@@ -1819,6 +1823,22 @@ export class App extends Reactor {
...
+     * Send a port absent message to federate with fed_ID, informing the
+     * remote federate that the current federate will not produce an event
+     * on this network port at the current logical time.
+     * 
+     * @param additional_delay The offset applied to the timestamp
+     *  using after. The additional delay will be greater or equal to zero
+     *  if an after is used on the connection. If no after is given in the
+     *  program, NEVER is passed.
+     * @param destFederatedID The fed ID of the receiving federate.
+     * @param destPortID The ID of the receiving port.
```
Comment (hokeun) [inline src/core/reactor.ts:1836]: I think we should update these comments as this function is just to be overridden by the subclass.

**Review by hokeun — APPROVED** [review]
Looks good to me. Just left minor suggestions.

**Comment by byeong-gil** [comment]
> Looks good to me. Just left minor suggestions.

These suggestions are all fixed in this [commit](https://github.com/lf-lang/reactor-ts/commit/20d16b96d5848ec9a721b1f19144870f7836cee8).

File: src/core/reactor.ts (hunk 5 above)
Comment (hokeun) [reply src/core/reactor.ts:1834]: I fixed this myself.

**Pull request merged**